kong/clustering/rpc/manager.lua

```diff
@@ -0,0 +1,365 @@
+    ngx_log(ngx_ERR, "[rpc] client did not provide node ID")
+    return ngx_exit(ngx.HTTP_CLOSE)
+  end
+
+  if content_encoding ~= "x-snappy-framed" then
+    ngx_log(ngx_ERR, "[rpc] client does use Snappy compressed frames")
+    return ngx_exit(ngx.HTTP_CLOSE)
+  end
+
+  if rpc_protocol ~= "kong.rpc.v1" then
+    ngx_log(ngx_ERR, "[rpc] unknown RPC protocol: " ..
+                     tostring(rpc_protocol) ..
+                     ", doesn't know how to communicate with client")
+    return ngx_exit(ngx.HTTP_CLOSE)
+  end
+
+  if not rpc_capabilities then
+    ngx_log(ngx_ERR, "[rpc] client did not provide capability list")
+    return ngx_exit(ngx.HTTP_CLOSE)
+  end
+
+  rpc_capabilities = cjson_decode(rpc_capabilities)
+  if not rpc_capabilities then
+    ngx_log(ngx_ERR, "[rpc] failed to decode client capability list")
+    return ngx_exit(ngx.HTTP_CLOSE)
+  end
+
+  local cert, err = validate_client_cert(self.conf, self.cluster_cert, ngx_var.ssl_client_raw_cert)
+  if not cert then
+    ngx_log(ngx_ERR, "[rpc] client's certificate failed validation: ", err)
+    return ngx_exit(ngx.HTTP_CLOSE)
+  end
+
+  ngx.header["X-Kong-RPC-Capabilities"] = cjson_encode(self.callbacks:get_capabilities_list())
+
+  local wb, err = server:new(WS_OPTS)
+  if not wb then
+    ngx_log(ngx_ERR, "[rpc] unable to establish WebSocket connection with client: ", err)
+    return ngx_exit(ngx.HTTP_CLOSE)
+  end
+
+  local s = socket.new(self, wb, node_id)
+  self:_add_socket(s, rpc_capabilities)
+
+  assert(s:start())
+  local res, err = s:join()
+  self:_remove_socket(s)
+
+  if not res then
+    ngx_log(ngx_ERR, "[rpc] RPC connection broken: ", err, " node_id: ", node_id)
+    return ngx_exit(ngx.ERROR)
+  end
+
+  return ngx_exit(ngx.OK)
+end
+
+
+function _M:connect(premature, node_id, host, path, cert, key)
+  if premature then
+    return
+  end
+
+  local uri = "wss://" .. host .. path
+
+  local opts = {
+    ssl_verify = true,
+    client_cert = cert,
+    client_priv_key = key,
+    protocols = "kong.rpc.v1",
+    headers = {
+      "X-Kong-Version: " .. KONG_VERSION,
+      "X-Kong-Node-Id: " .. self.node_id,
+      "X-Kong-Hostname: " .. kong.node.get_hostname(),
+      "X-Kong-RPC-Capabilities: " .. cjson_encode(self.callbacks:get_capabilities_list()),
+      "Content-Encoding: x-snappy-framed"
+    },
+  }
+
+  if self.conf.cluster_mtls == "shared" then
+    opts.server_name = "kong_clustering"
+
+  else
+    -- server_name will be set to the host if it is not explicitly defined here
+    if self.conf.cluster_server_name ~= "" then
+      opts.server_name = self.conf.cluster_server_name
+    end
+  end
+
+  local reconnection_delay = math.random(5, 10)
+
+  local c = assert(client:new(WS_OPTS))
+
+  local ok, err = c:connect(uri, opts)
+  if not ok then
+    ngx_log(ngx_ERR, "[rpc] unable to connect to peer: ", err)
+    goto err
+  end
+
+  do
+    local resp_headers = c:get_resp_headers()
+    -- FIXME: resp_headers should not be case sensitive
```

Comment:
It is easy to do, just add a `lower()` in lua-resty-websocket's patch? should we do it now?

Reply:
Could you fix it on the `lua-resty-websocket` branch? On the same branch with the `get_resp_headers()` PR.

Reply:
Add a `lower()` to key, see: https://github.com/chronolaw/lua-resty-websocket/pull/1/commits/e01e3bbb706a12e72b7b8e60a48ae329f30f5827.
Perhaps we should use `resty.http_headers` in lua-resty-http, is it a good idea?